### ISSUE TITLE
forcedDisplay and visibility="hidden"

### ISSUE THREAD

**Issue by nigelmegitt** [w3c/imsc#484]
Currently the specification for [`itts:forcedDisplay`](https://www.w3.org/TR/ttml-imsc1.1/#itts-forcedDisplay) in both IMSC 1.0.1 and IMSC 1.1 suggests in the introduction that it only applies to content for which the value of `tts:visibility="visible"` but the normative text does not actually make that distinction.

We have a use case for using `itts:forcedDisplay="true"` _and_ `tts:visibility="hidden"` for the same content: the idea is that when text is burned into video and has no equivalent sound an "invisible" caption is forced on to trigger reading by a screen reader, i.e. producing an audible presentation without producing "any visible rendering".

We want to do this in the short to medium term without having to advance to full adoption of `condition`, and this (direction of text to a screen reader) is not a candidate for the use of `tta:speak` which is itself not conditional on there being an active screen reader.

I don't think this alone is adequate to trigger the release of a new version of IMSC, however if one is being prepared for any reason, I would propose changing the opening sentence from:

> `itts:forcedDisplay` can be used to hide content whose computed value of `tts:visibility` is `"visible"` when the processor has been configured to do so via the application parameter `displayForcedOnlyMode`.

to 

> `itts:forcedDisplay` can be used to prevent visible rendering of content when the processor has been configured to do so via the application parameter `displayForcedOnlyMode`, regardless of the computed value of `tts:visibility`.

and add a note:

> Note: Applications that send presented text to a screen reader can be configured to do so even when normal display of captions is disabled by the user, without making any visible mark, by using `itts:forcedDisplay="true"` in conjunction with `tts:visibility="hidden"`. This can be used to cause on-screen text with no audio equivalent to be presented in an alternative medium for users who cannot read that text.

**Comment by palemieux** [comment]
> when text is burned into video and has no equivalent sound

Do you mean like the _high school_ example at https://www.w3.org/TR/ttml-imsc1.1/#forced-content , or something else?

**Comment by nigelmegitt** [comment]
@palemieux this is a different example. Here the visible text is _not_ provided as part of the TTML document, for rendering purposes, but is already present in the pixels of the video. A consequence of this antipattern is that the text is not accessible to screen reader users, therefore we want to send it in the TTML for non-visible presentation that _does_ get exposed to screen readers. Since it needs to be processed and "presented" even if subtitle presentation is apparently switched off, it is appropriate to use `itts:forcedDisplay="true"` and since we don't want it to result in any visible presentation, it also seems appropriate to use `tts:visibility="hidden"`.

**Comment by palemieux** [comment]
> Here the visible text is not provided as part of the TTML document

Hmmm. What do you mean by "here"? In the referenced annex, _Lycee_ *is* provided as part of the TTML document, as a translation of the _high school_ text burned in the video.

**Comment by nigelmegitt** [comment]
> What do you mean by "here"? 

By "here" I meant, in the new scenario described in this issue, which, incidentally, doesn't involve translation at all.

**Comment by palemieux** [comment]
My understanding of the use case:

- the listener is visually impaired, speaks English and uses a screen reader
- the video displays a scene with a a sign containing the text "danger: minefield" (in English)
- the audio does not contain any mention of "danger: minefield"
- the text "danger: minefield" should be sent to the user's screen reader during the scene

Is this the correct understanding?

**Comment by nigelmegitt** [comment]
Exactly right @palemieux .

**Comment by palemieux** [comment]
@cconcolato Input on how you would handle the use case at https://github.com/w3c/imsc/issues/484#issuecomment-519169604 ?

**Comment by palemieux** [comment]
@nigelmegitt It feels to me like this is new experience: it is neither captions for the hard-of-hearing (which are visible) nor audio description for the visually-impaired (which is heard). Is this something that the BBC does today, or plans to do in the future?

**Comment by nigelmegitt** [comment]
@palemieux possibly, it's something a bit like audio description for the visually-impaired, but not the complete experience. BBC currently publishes quite a large number of clips with this visual style of storytelling where the key information is presented as "narrative text", and this is seen as the quickest way to making them accessible for audiences who can't read the text.

**Comment by palemieux** [comment]
It see two distinct issues here:

(1) the introductory text is misleading: `itts:forcedDisplay` applies regardless of the computed value of `tts:visibility`. [ed.: I agree this should be corrected.]

(2) the proposed _Note_ at https://github.com/w3c/imsc/issues/484#issue-476894746 however hints at an application that goes beyond the intent of `itts:forcedDisplay`, and has implications for authors and processors, i.e. using `itts:forcedDisplay="true"` in conjunction with `tts:visibility="hidden"` and transmitting hidden text to screen readers. [ed.: this deserves a longer discussion, and likely a solution that goes beyond the  combination of `itts:forcedDisplay="true"` in conjunction with `tts:visibility="hidden"`.]

@nigelmegitt Do you wish to split the issue, or keep it as one?

**Comment by nigelmegitt** [comment]
@palemieux Happy to keep it as one for the time being, at least until we discuss it, noting we agreed to put this on the agenda for TPAC 2019 in our weekly call yesterday.

My reason is that I think a very reasonable outcome _might_ be that sending subtitle/caption text to a screen reader continues to be delegated to the player as an implementation-specific behaviour, as now, in which case only the changes I propose would need to be made. 

Of course another reasonable option is to be explicit about what player behaviour should be in relation to screen readers, in which case further edits would be needed and a separate issue would likely be opened.

As a data point, the BBC's Standard Media Player has sent subtitles and captions to screen readers for some time now, by default, using `aria-live="polite"` and `role="region"` on the subtitle container `div` in the HTML rendering.

**Comment by palemieux** [comment]
> Of course another reasonable option is to be explicit about what player behaviour 

@nigelmegitt I do not think it is necessary to be explicit about player behavior. It is however necessary to be explicit about the semantics of elements and attributes. I fairly certain that `tts:visibility="hidden"` Indicates that the element is not visible, perceivable, or interactive to any user -- to use the HTML definition. If the element is intended to be provided to a screen reader even though it is not rendered, then another attribute/element/track would be necessary.

**Comment by nigelmegitt** [comment]
Ah, I had assumed that complete removal from the tree would be done by setting `display: none;` or `tts:display="none"`. MDN [suggests](https://developer.mozilla.org/en-US/docs/Web/CSS/display#display_none) that a combination of properties can be used to make content visually invisible but available to screen readers, however in the linked page none of them includes setting `visibility`.

It's unfortunate that the HTML spec doesn't specify which CSS property is equivalent to making an element hidden in HTML. The fact that the note talks about setting the `display` property to override the `hidden` HTML attribute makes me think that the opposite is true too, that setting `display: none;` is equivalent to _setting_ the hidden attribute, but I don't think this is clear from that part of the specification anyway. 

In the ARIA documentation I see that in all the examples they match the value of the ARIA attribute to an equivalent value of the `visibility` property, for example in [`aria-hidden`](https://www.w3.org/TR/wai-aria/#aria-hidden) - this makes me think they are not intrinsically linked and that maybe it is possible to vary them independently. There's also an explicit mention of `display: none;` content _never_ being rendered regardless of anything else, which is what I would expect.

_If_ it is legitimate to use `visibility` in the way I proposed, then we should not add anything else; however if it is true that setting `visibility: hidden` forces the content to be hidden from the screen reader too, then I agree we may need to specify something special.

**Comment by palemieux** [comment]
> however if it is true that setting visibility: hidden forces the content to be hidden from the screen reader too, then I agree we may need to specify something special.

I suggest exploring approaches in that direction, e.g. a dedicated attribute.

Note that, in the case of forced text, it turns out that, in many scenarios, a dedicated track is created -- instead of using `itts:forcedDisplay`. This is necessary because the forced text shown on the screen changes when the non-forced text is displayed, e.g. there is less space for the forced text. So it might be the same here: a dedicated track might be needed for the "text-for-the-visually-impaired" use case.

**Comment by skynavga** [comment]
@nigelmegitt you probably know this, but `display` is very distinct from `visibility`, in both CSS and TTML: in both cases `display` controls whether (and how) content layout occurs, such that if it does occur, then it (generally) consumes layout space (in visual media), while `visibility` controls whether marks generated for that space are rendered (presented) or not; e.g.,

```
<span>Foo</span><span tts:display="none">Bar</span><span>Baz</span>
```

has no space between "Foo" and "Baz", while

```
<span>Foo</span><span tts:visibility="hidden">Bar</span><span>Baz</span>
```

has space (the width of "Bar") between "Foo" and "Baz";

the reason I'm making this comment is because I was not sure you made this distinction in your comment above;

**Comment by nigelmegitt** [comment]
Yes @skynavga I was aware of that; I did not make the distinction because I was only concerned with which property matches the HTML `@hidden` attribute, and because in the use case I have described, it isn't likely that there would be some inline hidden text in amongst non-hidden text. However, taking that semantic into account, the WHATWG specification of the [`@hidden`](https://html.spec.whatwg.org/multipage/interaction.html#the-hidden-attribute) attribute appears to fit more closely with `display: none;` than `visibility: hidden;`.

@palemieux I understand some content providers have made that choice; in this use case the forced text does not affect the non-forced text, so it is likely to be significantly easier to distribute a single TTML file and have the player switch presentation of the non-forced parts. If we can make it work, forced seems like a good fit.

**Comment by palemieux** [comment]
>  If we can make it work, forced seems like a good fit.

@nigelmegitt Sounds good. Can you propose an approach that is consistent with TTML, IMSC, ARAI, HTML, WCAG and MAUR?

**Comment by nigelmegitt** [comment]
> Can you propose an approach that is consistent with TTML, IMSC, ARAI, HTML, WCAG and MAUR?

@palemieux I _had_ thought that proposing `tts:visibility="hidden"` and having the player put the content into an ARIA live region would achieve that, so that's my baseline proposal until anyone tells me it doesn't work.

**Comment by palemieux** [comment]
>   had thought that proposing tts:visibility="hidden" and having the player put the content into an ARIA live region would achieve that, so that's my baseline proposal until anyone tells me it doesn't work.

The proposal as it stands does not work IMHO since it changes the semantics of `tts:visibility="hidden"`, which today means _not visible, perceivable, or interactive to any user_.

Can you propose an alternative/tweak?

**Comment by nigelmegitt** [comment]
> `tts:visibility="hidden"`, which today means _not visible, perceivable, or interactive to any user_.

@palemieux I haven't found any specification text to support this, looking at [TTML2](https://www.w3.org/TR/ttml2/#style-attribute-visibility), and the derivations at [XSL-FO](https://www.w3.org/TR/2006/REC-xsl11-20061205/#visibility) and [CSS](https://www.w3.org/TR/2011/REC-CSS2-20110607/visufx.html#visibility) - at the moment my understanding is that setting visibility to hidden is _not_ the same as setting the _hidden_ attribute in HTML, and only affects _visual_ presentation.

I've been experimenting with this in HTML and CSS and JS on Firefox and so far my experiments seem to bear this out(apologies I have to stop this soon and haven't got a clear shareable example yet):

* A `span` where `@hidden=true` has been set becomes visible when setting `display: inline;`
* A `span` where `@hidden=true` has been set does not become visible when setting `visibility: visible;`
* A `span` with `visibility: hidden` reports `@hidden = false`.
* A `span` with `display: none` reports `@hidden = false`.

In the first two cases, querying`@hidden` _after_ changing the CSS property results in `true` . So a `hidden` thing that is `visible` and has `display: inline` is _still rendered_.

The only consistent thing I've found is that setting `@hidden = false` on an element makes it disappear.

Right now it is not clear to me how this is supposed to work, but I haven't managed to work my way around to understanding how the `@hidden` attribute's [description](https://html.spec.whatwg.org/multipage/interaction.html#the-hidden-attribute) relates directly to anything in CSS. One thing I notice is that it includes the text "User agents should not render elements that have the hidden attribute specified." - interesting "should" there!

**Comment by palemieux** [comment]
@nigelmegitt Have you considered defining an explicit attribute, akin to `forcedDisplay`, but for screen readers? Any downside? Would a potential next step be to seek feedback from CSS and the accessibility WG?

**Comment by nigelmegitt** [comment]
> Have you considered defining an explicit attribute

@palemieux I'd consider a new attribute if I felt certain that there is no existing way to get the same effect; I'm not really convinced of that yet, but it certainly wouldn't harm to get feedback from those groups. I'll ask them.

**Comment by nigelmegitt** [comment]
Further evidence: https://github.com/w3c/csswg-drafts/issues/560 and https://developer.mozilla.org/en-US/docs/Web/CSS/visibility#Accessibility_concerns suggest that as you stated @palemieux setting `visibility: hidden;` also should hide the content from the screen reader, and that other people have asked for this too.

**Comment by palemieux** [comment]
@nigelmegitt After reviewing the thread above, I do not see a concrete action for IMSC 1.3. Do you have a different opinion?

**Comment by nigelmegitt** [comment]
I think we need to reach out to the CSS WG to see what their a11y TF came up with.

Also, @palemieux have you changed your view (item 1 at https://github.com/w3c/imsc/issues/484#issuecomment-519661953) that the `itts:forcedDisplay` applies regardless of `tts:visibility`?
